lib/rouge/lexers/ttcn3.rb

```diff
@@ -0,0 +1,129 @@
+# -*- coding: utf-8 -*- #
+# frozen_string_literal: true
+
+module Rouge
+  module Lexers
+    class TTCN3 < RegexLexer
+      title "TTCN3"
+      desc "The TTCN3 programming language (ttcn-3.org)"
+
+      tag 'ttcn3'
+      filenames '*.ttcn', '*.ttcn3'
+      mimetypes 'text/x-ttcn3', 'text/x-ttcn'
+
+      def self.keywords
+        @keywords = %w(
+          module import group type port component signature external
+          execute const template function altstep testcase var timer if
+          else select case for while do label goto start stop return
+          break int2char int2unichar int2bit int2enum int2hex int2oct
+          int2str int2float float2int char2int char2oct unichar2int
+          unichar2oct bit2int bit2hex bit2oct bit2str hex2int hex2bit
+          hex2oct hex2str oct2int oct2bit oct2hex oct2str oct2char
+          oct2unichar str2int str2hex str2oct str2float enum2int
+          any2unistr lengthof sizeof ispresent ischosen isvalue isbound
+          istemplatekind regexp substr replace encvalue decvalue
+          encvalue_unichar decvalue_unichar encvalue_o decvalue_o
+          get_stringencoding remove_bom rnd hostid send receive
+          setverdict
+        )
+      end
+
+      def self.reserved
+        @reserved = %w(
+          all alt apply assert at configuration conjunct const control
+          delta deterministic disjunct duration fail finished fuzzy
+          history implies inconc inv lazy mod mode notinv now omit
+          onentry onexit par pass prev realtime seq setstate static
+          stepsize stream timestamp until values wait
+        )
+      end
+
+      def self.types
+        @types = %w(
+          anytype address boolean bitstring bytestring charstring
+          component enumerated float integer hexstring octetstring port
+          record set of union universal
+        )
+      end
+
+      # optional comment or whitespace
+      ws = %r((?:\s|//.*?\n|/[*].*?[*]/)+)
+      id = /[a-zA-Z_][a-zA-Z0-9_]*/
+      const_name = /[A-Z][A-Z0-9_]*\b/
+      module_name = /[A-Z][a-zA-Z0-9]*\b/
+
+      state :root do
+        rule %r/[^\S\n]+/, Text
+
+        rule %r(//.*), Comment::Single
+        rule %r(/\*.*?\*/)m, Comment::Multiline
+
+        # keywords: go before method names to avoid lexing "throw new XYZ"
+        # as a method signature
+        rule %r(
```

Comment:
Could you explain this rule a little? I only gave the TTCN3 documentation a cursory glance but it looked to me like functions are prefaced with the `function` keyword. I didn't see anything that matched how this appears to work.

Reply:
Wrong copy past when I staterted to create the TTCN-3 lexer. Removed

Reply:
When you delegate lexing of the return arguments back to the TTCN-3 lexer itself, all it will do is go through and match the arguments against the other rules in this lexer. Given that, why match them with this rule initially? Can't you identify a function/method merely by the fact that it has an opening parenthesis? If you can drop the arguments, you can just have something like:

```ruby
         rule %r(
           ([a-zA-Z_][\w]*)                    # method name
           (\s*)(\()                           # signature start
         )mx do |m|
           groups Name::Function, Text, Punctuation
         end
```

Reply:
Hello,
My apologies for the delay.
I did a full review of the TTCN-3 rules. I think it looks better now.